sw/otbn/crypto/div.s

```diff
@@ -0,0 +1,369 @@
+ * extra DMEM space allocated for it.
+ *
+ * Flags have no meaning beyond the scope of this subroutine.
+ *
+ * @param[in]      x3: dptr_a, pointer to input a in DMEM
+ * @param[in]      x9: n, number of 256-bit limbs
+ * @param[in]     x23: 23, constant
+ * @param[in]     x24: 24, constant
+ * @param[in]     w31: all-zero
+ * @param[out]    w23: highest limb of original a value
+ *
+ * clobbered registers: x3, w23, w24
+ * clobbered flag groups: none
+ */
+bignum_lshift256:
+  /* Loop invariants for iteration i:
+       x3 = dptr_a + i*32
+       w23 = if i == 0 then 0 else a[i-1]
+       dmem[dptr_a..dptr_a+i*32] = (a << 256) mod 2^(i*32) */
+  bn.mov    w23, w31
+  loop      x9, 3
+    /* w24 <= a[i] */
+    bn.lid    x24, 0(x3)
+    /* a[i] <= w23 = a[i-1] */
+    bn.sid    x23, 0(x3++)
+    /* w23 <= w24 */
+    bn.mov    w23, w24
+
+  ret
+
+/**
+ * Conditionally subtract a shifted bignum from another bignum.
+ *
+ * Returns r' = if (r > y << (i*256)) then (r - y << (i*256)) else r.
+ *
+ * Updates r in-place.
+ *
+ * This is a specialized helper function for `div`. It expects its inputs in
+ * specific forms that come from the constraints of that function. In
+ * particular, the second operand y has n+1 limbs, one extra compared to r,
+ * with the uppermost stored in a register.
+ *
+ * For the first i limbs, the shifted y is 0 so r will always remain unchanged.
+ * For the next (n-i) iterations, we subtract corresponding limbs.  For the
+ * last i iterations, we would need to subtract the remaining limbs of y from
+ * 0; however, since we would never use this result if it were nonzero, we
+ * simply compute the borrow rather than a full subtraction. Finally, we check
+ * the borrow at the end of the computation and, if the subtraction
+ * underflowed, we add back the limbs of y that we subtracted to leave r
+ * unchanged.
+ *
+ * Algorithmically, this looks something like:
+ *   borrow = 0
+ *   for k=0..n-1:
+ *     if k+i < n:
+ *       r[k+i], borrow = subb(r[k+i], y[k], borrow)
+ *     else:
+ *       borrow &= (y[k] == 0)
+ *   mask = borrow ? 2^256 - 1 : 0
+ *   carry = 0
+ *   for k=0..n-i-1:
+ *     r[k+i], carry = addc(r[k+i], y[k] & mask, carry)
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * @param[in]    x8: i, number of limbs to shift y (i < n)
+ * @param[in]    x9: n, number of 256-bit limbs for operands r and y
+ * @param[in]   x10: dptr_r, pointer to first operand r in dmem (n*32 bytes)
+ * @param[in]   x11: dptr_y, pointer to second operand y in dmem (n*32 bytes)
+ * @param[in]   x24: 24, constant
+ * @param[in]   x25: 25, constant
+ * @param[in]   w27: y_upper, highest limb of y
+ * @param[in]   w31: all-zero
+ * @param[out]  w23: mask, 0 if the subtraction was performed and -1 otherwise
+ * @param[out] dmem[dptr_r..dptr_r+n*32]: result r'
+ *
+ * clobbered registers: x2 to x5, w23 to w25
+ * clobbered flag groups: FG0
+ */
+cond_sub_shifted:
```

Comment:
Nit: I think the comment header for this function doesn't mention that it's constant time.

Reply:
Thanks for the catch! I'll add this line in the next PR where I add tests, just to avoid waiting for CI again.